statistics/handle/handle.go

```diff
@@ -962,3 +962,150 @@ func (h *Handle) ReloadExtendedStatistics() error {
+			continue
+		}
+		item = h.fillExtendedStatsItemVals(item, cols, collectors)
+		if item != nil {
+			statsColl.Stats[key] = item
+		}
+	}
+	if len(statsColl.Stats) == 0 {
+		return nil, nil
+	}
+	return statsColl, nil
+}
+
+func (h *Handle) fillExtendedStatsItemVals(item *statistics.ExtendedStatsItem, cols []*model.ColumnInfo, collectors []*statistics.SampleCollector) *statistics.ExtendedStatsItem {
+	switch item.Tp {
+	case ast.StatsTypeCardinality, ast.StatsTypeDependency:
+		return nil
+	case ast.StatsTypeCorrelation:
+		return h.fillExtStatsCorrVals(item, cols, collectors)
+	}
+	return nil
+}
+
+func (h *Handle) fillExtStatsCorrVals(item *statistics.ExtendedStatsItem, cols []*model.ColumnInfo, collectors []*statistics.SampleCollector) *statistics.ExtendedStatsItem {
+	colOffsets := make([]int, 0, 2)
+	for _, id := range item.ColIDs {
+		for i, col := range cols {
+			if col.ID == id {
+				colOffsets = append(colOffsets, i)
+				break
+			}
+		}
+	}
+	if len(colOffsets) != 2 {
+		return nil
+	}
+	// samplesX and samplesY are in order of handle, i.e, their SampleItem.Ordinals are in order.
+	samplesX := collectors[colOffsets[0]].Samples
+	// We would modify Ordinal of samplesY, so we make a deep copy.
+	samplesY := statistics.CopySampleItems(collectors[colOffsets[1]].Samples)
+	sampleNum := len(samplesX)
+	if sampleNum == 1 {
+		item.ScalarVals = float64(1)
+		return item
+	}
+	h.mu.Lock()
+	sc := h.mu.ctx.GetSessionVars().StmtCtx
+	h.mu.Unlock()
+	var err error
+	samplesX, err = statistics.SortSampleItems(sc, samplesX)
+	if err != nil {
+		return nil
+	}
+	samplesYInXOrder := make([]*statistics.SampleItem, sampleNum)
+	for i, itemX := range samplesX {
+		itemY := samplesY[itemX.Ordinal]
+		itemY.Ordinal = i
+		samplesYInXOrder[i] = itemY
+	}
+	samplesYInYOrder, err := statistics.SortSampleItems(sc, samplesYInXOrder)
+	if err != nil {
+		return nil
+	}
+	var corrXYSum float64
+	for i := 1; i < sampleNum; i++ {
+		corrXYSum += float64(i) * float64(samplesYInYOrder[i].Ordinal)
+	}
+	// X means the ordinal of the item in original sequence, Y means the oridnal of the item in the
+	// sorted sequence, we know that X and Y value sets are both:
+	// 0, 1, ..., sampleNum-1
+	// we can simply compute sum(X) = sum(Y) =
+	//    (sampleNum-1)*sampleNum / 2
+	// and sum(X^2) = sum(Y^2) =
+	//    (sampleNum-1)*sampleNum*(2*sampleNum-1) / 6
+	// We use "Pearson correlation coefficient" to compute the order correlation of columns,
+	// the formula is based on https://en.wikipedia.org/wiki/Pearson_correlation_coefficient.
+	// Note that (itemsCount*corrX2Sum - corrXSum*corrXSum) would never be zero when sampleNum is larger than 1.
+	itemsCount := float64(sampleNum)
+	corrXSum := (itemsCount - 1) * itemsCount / 2.0
+	corrX2Sum := (itemsCount - 1) * itemsCount * (2*itemsCount - 1) / 6.0
+	item.ScalarVals = (itemsCount*corrXYSum - corrXSum*corrXSum) / (itemsCount*corrX2Sum - corrXSum*corrXSum)
```

Comment:
There is the same logic about calculating correlation values in `BuildColumnHist`, so should we extract this logic as a separate function?

Reply:
This logic needs an input `corrXYSum`, extracting this logic would separate the computing of `corrXYSum` into another function, which looks pretty weird, so I didn't extract then.